cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1853,24 +1853,46 @@ func getNonUEFISystemDisk(fallbacklabel string) (string, error) {
 // kernel came from cannot be determined, then it will fallback to mounting via
 // the specified disk label.
 func mountNonDataPartitionMatchingKernelDisk(dir, fallbacklabel string, opts *systemdMountOptions) error {
-	partuuid, err := bootFindPartitionUUIDForBootedKernelDisk()
 	var partSrc string
-	if err == nil {
-		// TODO: the by-partuuid is only available on gpt disks, on mbr we need
-		//       to use by-uuid or by-id
-		partSrc = filepath.Join("/dev/disk/by-partuuid", partuuid)
-	} else {
-		partSrc, err = getNonUEFISystemDisk(fallbacklabel)
+
+	if osutil.FileExists(filepath.Join(dirs.GlobalRootDir, "/dev/disk/snapd/disk")) {
+		disk, err := disks.DiskFromDeviceName("/dev/disk/snapd/disk")
 		if err != nil {
 			return err
 		}
-	}
+		partuuid, err := bootFindPartitionUUIDForBootedKernelDisk()
+		if err == nil {
+			partition, err := disk.FindMatchingPartitionWithPartUUID(partuuid)
+			if err != nil {
+				return err
+			}
+			partSrc = partition.KernelDeviceNode
+		} else {
+			partition, err := disk.FindMatchingPartitionWithFsLabel(fallbacklabel)
+			if err != nil {
+				return err
+			}
+			partSrc = partition.KernelDeviceNode
+		}
```

Comment:
We could instead look at the `/dev/disk/snapd/ubuntu-{seed,boot}` symlinks now, right? Not saying that we should, but maybe adding a comment about that would be interesting.

Reply:
Yes, that is the plan. But the PR is already big.

Reply:
We can just continue to depend on disk. There either partition should be there.

The only problem is that we have a lot of places where we resolve partitions, and it will be a lot to change.

Reply:
Ah also, we need to have a way to resolve still when the rules do not exist on that initrd.

Reply:
Would race conditions be an issue? The symlinks to the partitions are created after the one for the disk, so I was wondering. Anyway, we will get to this if we do the change.